pkg/csi_driver/controller.go

```diff
@@ -110,6 +137,28 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
 	return &csi.CreateVolumeResponse{Volume: fileInstanceToCSIVolume(filer, modeInstance)}, nil
 }
 
+// getUnreservedIPBlock returns the available IP in the cidr. It is synchronized in order to avoid data race while getting unreserved IP range in a CIDR
+func (s *controllerServer) reserveIPV4Range(ctx context.Context, filer *file.ServiceInstance, cidr string) (string, error) {
+	instances, err := s.config.fileService.ListInstances(ctx, filer)
+	if err != nil {
+		return "", status.Error(codes.Aborted, err.Error())
+	}
+
+	// Initialize an empty reserved list. It will be populated with all the reservedIPRanges obtained from the active instances
+	reservedIPRangesActiveInstances := make(map[string]bool)
+	for _, instance := range instances {
+		reservedIPRangesActiveInstances[instance.Network.ReservedIpRange] = true
+	}
+
+	// Pass the reservedIPRanges obtained from ListInstances call. We have the IPRanges pending reservation in the IPAllocator.hold list
+	// The GetUnreservedIPBlock would merge these two lists to form a reserved list
```

Comment:
This comment is an implementation detail and would be better suited for inside the util library